src/main/java/com/redhat/foreman/DisposableImpl.java

```diff
@@ -0,0 +1,40 @@
+package com.redhat.foreman;
+
+import org.apache.log4j.Logger;
+import org.jenkinsci.plugins.resourcedisposer.Disposable;
+
+import javax.annotation.Nonnull;
+
+/**
+ * Created by shebert on 27/09/16.
+ */
+public class DisposableImpl implements Disposable {
+
+    private final String cloudName;
+    private final String name;
+    private static final Logger LOGGER = Logger.getLogger(DisposableImpl.class);
+
+    public DisposableImpl(String cloudName, String name) {
+        this.cloudName = cloudName;
+        this.name = name;
+    }
+
+    @Nonnull
+    @Override
+    public State dispose() throws Exception {
+        ForemanSharedNodeCloud cloud = ForemanSharedNodeCloud.getByName(cloudName);
+        if (cloud != null) {
+            cloud.getForemanAPI().release(name);
```

Comment:
> In the event that the resource is freed externally, the release method first checks to make sure it is reserved by the Jenkins instance itself using getReserveReason().

Cool, that addresses my concern.

> So what is you are saying is that in dispose(), I should check if Foreman is available first thru another API call. If not available, then return with State.TO_DISPOSE?

It is preferable to throw as it gives admins a chance to know *why* it was not disposed.